README
This is Perl module Jonk.

INSTALLATION

Jonk installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan Jonk

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

DOCUMENTATION

Jonk documentation is available as in POD. So you can do:

    % perldoc Jonk

to read the documentation online with your favorite pager.

Atsushi Kobayashi


Changes
Revision history for Perl extension Jonk

0.10_02
   2011-02-08
        - refactoring code.
        - can set (de)serializer each func or global.
        - can set priority by $job->failed.
        - enqueue_time change format to unixtime.

0.10_01
   2011-01-24
        [ INCOMPATIBLE CHANGE!! ]
        - many api move to Jonk.
        - Jonk::Client, Jonk::Worker is removed.
        - create Jonk::Job.

0.06
   2010-12-16
        - FIXED : some pod.

0.05
   2010-12-01
        - FIXED : no use Pod::Simple for testing.

0.04
   2010-11-29
        - FIXED : requires 'DBI' for Makefile.PL.

0.03
   2010-11-28
        - FIXED : test use DBD::SQLite.

0.02
   2010-11-26
        - ADD   : more document.
        - FIXED : Client->_insert_id on PostgreSQL (thanks fujiwara.)

0.01
   2010-11-25
        - original version


t/Utils.pm
package t::Utils;
use strict;
use warnings;
use DBI;
use Test::More;
use File::Temp qw(tempdir);

BEGIN {
  eval "use DBD::SQLite";
  plan skip_all => 'needs DBD::SQLite for testing' if $@;
}

sub import {
    strict->import;
    warnings->import;
}

sub _get_schema {
    my $dbh   = shift;
    my $table = shift || "job";

    my $driver = $dbh->{Driver}{Name};
    my $sql;
    if ( $driver eq 'mysql' ) {
        $sql = _read_pod('MySQL');
    } elsif ( $driver eq 'Pg' ) {
        $sql = _read_pod('PostgreSQL');
    } elsif ( $driver eq 'SQLite' ) {
        $sql = _read_pod('SQLite');
    } else {
        Carp::croak "this driver unsupport: $driver";
    }
    $sql =~ s/(CREATE\s+TABLE\s+)job/CREATE TABLE $table/;
    $sql;
}

sub _read_pod {
    my $type = shift;

    open my $fh, '<', './lib/Jonk.pm';
    my $do_read     = 0;
    my $read_schema = 0;
    my $schema='';
    while (<$fh>) {
        my $line = $_;
        if ($line =~ /=head1 SCHEMA/) {
            $do_read=1;
        }
        if ($do_read) {
            if ($line =~ /$type/) {
                $read_schema = 1;
                next;
            }
            if ($read_schema && $line =~ /^=head/) {
                $read_schema = 0;
                last;
            }
            if ($read_schema) {
                $schema .= $line;
                next;
            }
        }
    }
    $schema;
}

my $db;
my $_setup=0;
sub setup {
    my ($class, $table) = @_;

    $db ||= do {
        my $tempdir = tempdir(CLEANUP => 1);
        File::Spec->catfile($tempdir, 'jonk_test.db');
    };

    my $dbh = DBI->connect('dbi:SQLite:'.$db);
    $_setup ? do {1} : do {$dbh->do(_get_schema($dbh, $table)); $_setup=1};
    return $dbh;
}

sub cleanup {
    my ($class, $dbh, $table) = @_;
    $table ||= "job";
    $dbh->do("DROP TABLE $table");
    $_setup=0;
}

1;



t/start_mysqld.pl
#! perl
use t::Utils;
use JSON::XS;
use DBI;

$SIG{INT} = sub { CORE::exit 1 };
$mysqld = t::Utils->setup;
$ENV{TEST_MYSQLD} = encode_json +{ %$mysqld };

my $dbh = DBI->connect($mysqld->dsn(dbname => 'test'));
$dbh->do(
q{
CREATE TABLE job (
    id           int(10) unsigned NOT NULL auto_increment,
    func         varchar(255)     NOT NULL,
    arg          MEDIUMBLOB,
    enqueue_time DATETIME         NOT NULL,
    primary key ( id )
)
}
);



bench/bench.pl
#! /usr/bin/perl
use strict;
use warnings;
use Benchmark qw/countit timethese timeit timestr/;
use Jonk::Client;
use Jonk::Worker;
use DBI;
use Parallel::ForkManager;

my $db = DBI->connect('dbi:mysql:jonk','root','');
$db->do(q{DROP TABLE job});
$db->do(q{
    CREATE TABLE job (
        id           int(10) unsigned NOT NULL auto_increment,
        func         varchar(255)     NOT NULL,
        arg          MEDIUMBLOB,
        enqueue_time DATETIME         NOT NULL,
        primary key ( id )
    )
});
my $pm = Parallel::ForkManager->new(50);

my ( $sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst ) = localtime(time);
my $time = sprintf('%04d-%02d-%02d %02d:%02d:%02d', $year + 1900, $mon + 1, $mday, $hour, $min, $sec);

timethese(10, {
    'enqueue' => sub {
        for my $i (1 .. 10000) {
            my $pid = $pm->start and next;

                my $dbh = DBI->connect('dbi:mysql:jonk','root','');
                my $jonk = Jonk::Client->new($dbh);
                my $job_id = $jonk->enqueue('MyWorker3',+{arg => 'args_'.$i, time => $time});

            $pm->finish;
        }
        $pm->wait_all_children;
    },
});

timethese(10, {
    'dequeue' => sub {
        for my $i (1 .. 10000) {
            my $pid = $pm->start and next;

            my $dbh = DBI->connect('dbi:mysql:jonk','root','');
            my $jonk = Jonk::Worker->new($dbh,{functions => [qw/MyWorker2/]});
            my $job = $jonk->dequeue;

            $pm->finish;
        }
        $pm->wait_all_children;
    },
});

__END__
Benchmark: timing 5 iterations of enqueue...
   enqueue: 24 wallclock secs ( 0.67 usr  2.67 sys + 11.93 cusr 21.04 csys = 36.31 CPU) @  0.14/s (n=5)
Benchmark: timing 5 iterations of dequeue...
   dequeue: 55 wallclock secs ( 0.74 usr  3.70 sys + 12.73 cusr 28.16 csys = 45.33 CPU) @  0.11/s (n=5)



lib/Jonk/Job.pm
package Jonk::Job;
use strict;
use warnings;
use Carp ();

sub new {
    my ($class, $jonk, $job) =@_;
    bless {
        job        => $job,
        _jonk      => $jonk,
        _completed => 0,
        _failed    => 0,
        _aborted   => 0,
    }, $class;
}

sub completed {
    my $self = shift;

    Carp::croak 'job is already failed.'  if $self->is_failed;
    Carp::croak 'job is already aborted.' if $self->is_aborted;

    $self->{_completed} = 1;
    $self->{_jonk}->_delete($self->id);
}

sub failed {
    my ($self, $opt) = @_;

    Carp::croak 'job is already complated.' if $self->is_completed;
    Carp::croak 'job is already aborted.'   if $self->is_aborted;

    $self->{_failed} = 1;
    $self->{_jonk}->_failed(
        $self->id => $opt
    );
}

sub aborted {
    my $self = shift;

    Carp::croak 'job is already complated.' if $self->is_completed;
    Carp::croak 'job is already failed.'    if $self->is_failed;

    $self->{_aborted} = 1;
    $self->{_jonk}->_delete($self->id);
}

sub id            { $_[0]->{job}->{id}            }
sub func          { $_[0]->{job}->{func}          }
sub raw_arg       { $_[0]->{job}->{arg}           }
sub arg {
    my $self = shift;
    $self->{_jonk}->{functions}->{$self->func}->{deserializer}->($self->{job}->{arg});
}
sub enqueue_time  { $_[0]->{job}->{enqueue_time}  }
sub grabbed_until { $_[0]->{job}->{grabbed_until} }
sub retry_cnt     { $_[0]->{job}->{retry_cnt}     }
sub run_after     { $_[0]->{job}->{run_after}     }
sub priority      { $_[0]->{job}->{priority}      }

sub is_completed { $_[0]->{_completed} }
sub is_failed    { $_[0]->{_failed}    }
sub is_aborted   { $_[0]->{_aborted}   }

sub DESTROY {
    my $self = shift;
    unless ($self->is_completed or $self->is_aborted or $self->is_failed) {
        Carp::cluck "job is not (complete|fail|abor)ed. this job auto failed.";
        $self->failed;
    }
}

1;

__END__

__END__

=head1 NAME

Jonk::Job - Jonk job class.

=head1 SYNOPSIS

    my $job = $jonk->lookup_job($job_id);
    $job->arg;
    $job->completed; # complete and delete job
    $job->failed;    # failed and update job for retry
    $job->aborted;   # failed and delete job

=head1 METHODS

=head2 $job->id           

get job id.

=head2 $job->func         

get job function name

=head2 $job->arg.         

get job argument.

=head2 $job->enqueue_time 

get job enqueued time.

=head2 $job->grabbed_until

get job grabbed until time.

=head2 $job->retry_cnt    

get job retried count

=head2 $job->run_after    

get job run after time

=head2 $job->priority     

get job priority num

=head2 $job->completed

completed job.

and delete job data from database.

=head2 $job->failed([$options]);

failed job.
set error message and delete job data from database.

=over 4

=item * $options->{retry_delay}

job retry delay sec.

Default 60 sec.

=item * $options->{priority}

job priority number.

Default 0.

=back

=head2 $job->aborted

aborted job.
set error message and delete job data from database.

=head2 $job->is_completed

determinate job completed or not.

when $job->completed is called, it is set true.

=head2 $job->is_failed

determinate job failed or not.

when $job->failed is called, it is set true.

=head2 $job->is_aborted

determinate job aborted or not.

when $job->aborted is called, it is set true.

=cut



lib/Jonk/Client.pm
package Jonk::Client;
use strict;
use warnings;
use Carp ();
use Try::Tiny;

sub new {
    my ($class, $dbh, $opts) = @_;

    unless ($dbh) {
        Carp::croak('missing job queue database handle.');
    }
    my $table_name = $opts->{table_name} || "job";
    bless {
        dbh           => $dbh,
        enqueue_query => sprintf('INSERT INTO %s (func, arg, enqueue_time) VALUES (?,?,?)', $table_name),
        enqueue_time_callback => ($opts->{enqueue_time_callback}||sub{
            my ( $sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst ) = localtime(time);
            return sprintf('%04d-%02d-%02d %02d:%02d:%02d', $year + 1900, $mon + 1, $mday, $hour, $min, $sec);
        }),
        table_name    => $table_name,
        _errstr       => undef,
    }, $class;
}

sub errstr {$_[0]->{_errstr}}

sub enqueue {
    my ($self, $func, $arg) = @_;

    my $job_id;
    try {
        $self->{_errstr} = undef;
        local $self->{dbh}->{RaiseError} = 1;
        local $self->{dbh}->{PrintError} = 0;

        my $sth = $self->{dbh}->prepare_cached($self->{enqueue_query});
        $sth->bind_param(1, $func);
        $sth->bind_param(2, $arg, _bind_param_attr($self->{dbh}));
        $sth->bind_param(3, $self->{enqueue_time_callback}->());
        $sth->execute();

        $job_id = $self->_insert_id($self->{dbh});
        $sth->finish;
    } catch {
        $self->{_errstr} = "can't enqueue for job queue database: $_"
    };

    $job_id;
}

sub _insert_id {
    my $self = shift;
    my $dbh  = shift;

    my $driver = $dbh->{Driver}{Name};
    if ( $driver eq 'mysql' ) {
        return $dbh->{mysql_insertid};
    } elsif ( $driver eq 'Pg' ) {
        return $dbh->last_insert_id( undef, undef, undef, undef,{ sequence => join( '_', $self->{table_name}, 'id', 'seq' ) } );
    } elsif ( $driver eq 'SQLite' ) {
        return $dbh->func('last_insert_rowid');
    } else {
        Carp::croak "Don't know how to get last insert id for $driver";
    }
}

sub _bind_param_attr {
    my $dbh = shift;

    my $driver = $dbh->{Driver}{Name};
    if ( $driver eq 'Pg' ) {
        return { pg_type => DBD::Pg::PG_BYTEA() };
    } elsif ( $driver eq 'SQLite' ) {
        return DBI::SQL_BLOB();
    }
    return;
}

1;
__END__

=head1 NAME

Jonk::Client - job enqueue client class.

=head1 SYNOPSIS

    use DBI; 
    use Jonk::Client;
    use DateTime;
    
    my $dbh = DBI->connect(...);
    my $jonk = Jonk::Client->new($dbh, 
        {
            table_name => 'jonk_job',
            enqueue_time_callback => sub {
                DateTime->now->strftime('%Y-%m-%d %H:%M:%S');
            },
        }
    );

=head1 METHODS

=head2 my $jonk = Jonk::Client->new($dbh, [$options]);

Creates a new Jonk object, and returns the object.

$option is an optional settings.

=over 4

=item * $dbh

$dbh is database handle.

=item * $options->{table_name}

specific job table name.

Default job table name is `job`.

=item * $options->{enqueue_time_callback}

specific enqueue_time creation callback.

Default local time create.

=back

=head2 my $job_id = $jonk->enqueue($func, $arg);

enqueue a job to a database.
returns job.id.

=over 4

=item * $func

specific your worker funcname.

=item * $arg

job argument data.

serialize is not done in Jonk. 

Please pass data that does serialize if it is necessary. 

=back

=head2 my $errstr = $jonk->errstr;

get most recent error infomation.

=head1 ERROR HANDLING

  my $job_id = $jonk->enqueue('func','arg');
  if ($jonk->errstr) {
      die $jonk->errstr;
  }

=cut



lib/Jonk/Worker.pm
package Jonk::Worker;
use strict;
use warnings;
use Carp ();
use Try::Tiny;

sub new {
    my ($class, $dbh, $opts) = @_;

    unless ($dbh) {
        Carp::croak('missing job queue database handle.');
    }

    my $table_name = ($opts->{table_name}||'job');
    bless {
        dbh              => $dbh,
        lookup_job_query => sprintf('SELECT * FROM %s WHERE id = ?', $table_name),
        find_job_query   => sprintf('SELECT * FROM %s WHERE func IN (%s) ORDER BY id LIMIT %s',
                             $table_name,
                             join(', ', map { "'$_'" } @{$opts->{functions}}),
                             ($opts->{job_find_size}||50),
                         ),
        dequeue_query    => sprintf('DELETE FROM %s WHERE id = ?', ($opts->{table_name}||'job')),
        _errstr          => undef,
    }, $class;
}

sub errstr {$_[0]->{_errstr}}

sub dequeue {
    my ($self, $job_id) = @_;

    my $job;
    try {
        $self->{_errstr} = undef;
        local $self->{dbh}->{RaiseError} = 1;
        local $self->{dbh}->{PrintError} = 0;

        my $sth;
        if ($job_id) {
            $sth = $self->{dbh}->prepare_cached($self->{lookup_job_query});
            $sth->execute($job_id);
        } else {
            $sth = $self->{dbh}->prepare_cached($self->{find_job_query});
            $sth->execute();
        }

        while (my $row = $sth->fetchrow_hashref) {
            my $del_sth = $self->{dbh}->prepare_cached($self->{dequeue_query});
            $del_sth->execute($row->{id});
            $del_sth->finish;

            if ($del_sth->rows) {
                $job = $row;
                last;
            }
        }

        $sth->finish;
    } catch {
        $self->{_errstr} = "can't get job from job queue database: $_";
    };

    $job;
}

1;
__END__

=head1 NAME

Jonk::Worker - get a job data class.

=head1 SYNOPSIS

    use DBI; 
    use Jonk::Worker;
    
    my $dbh = DBI->connect(...);
    my $jonk = Jonk::Worker->new($dbh, {functions => ['MyWorker']});
    my $job = $jonk->dequeue;
    print $job->{func}; # MyWorker
    print $job->{arg};  # arg

=head1 METHODS

=head2 my $jonk = Jonk::Worker->new($dbh, [$options]);

Creates a new Jonk object, and returns the object.

$options is an optional settings.

=over 4

=item * $dbh

$dbh is database handle.

=item * $options->{functions}

Key word of job which this Jonk instance looks for.

=item * $options->{table_name}

specific job table name.

Default job table name is `job`.

=item * $options->{job_find_size}

specific lookup job record size.

Default 50.

=back

=head2 my $job_hash_ref = $jonk->dequeue([$job_id]);

dequeue a job from a database.

returns job hashref data.

Please do deserialize if it is necessary. 

$job_id is optional argument.

=over 4

=item * $job_id (optional)

lookup specific $job_id's job.

=back

=head2 $jonk->errstr;

get most recent error infomation.

=head1 ERROR HANDLING

  my $job = $jonk->dequeue;
  if ($jonk->errstr) {
      die $jonk->errstr;
  }

=cut



lib/Jonk.pm
package Jonk;
use strict;
use warnings;
use Jonk::Job;
use Try::Tiny;
use Carp ();

our $VERSION = '0.10_02';

sub new {
    my ($class, $dbh, $opts) = @_;

    unless ($dbh) {
        Carp::croak('missing job queue database handle.');
    }

    my $functions  = _parse_functions($opts);
    my $table_name = $opts->{table_name} || 'job';
    my $driver     = _verify_driver($dbh);

    bless {
        dbh           => $dbh,
        table_name    => $table_name,
        functions     => $functions,
        driver        => $driver,
        has_func      => scalar(keys %{$functions}) ? 1 : 0,

        _errstr       => undef,

        insert_query => sprintf(
            'INSERT INTO %s (func, arg, enqueue_time, grabbed_until, run_after, retry_cnt, priority) VALUES (?,?,?,0,?,0,?)'
            ,$table_name
        ),

        grab_query => sprintf('UPDATE %s SET grabbed_until = ? WHERE id = ? AND grabbed_until = ?', $table_name),

        lookup_job_query => sprintf('SELECT * FROM %s WHERE id = ? AND grabbed_until <= ? AND run_after <= ?', $table_name),

        find_job_query => sprintf(
            'SELECT * FROM %s WHERE func IN (%s) AND grabbed_until <= ? AND run_after <= ? ORDER BY priority DESC LIMIT %s',
            $table_name, (join(', ', map { "'$_'" } keys %{$functions})), ($opts->{job_find_size} || 50)
        ),

        delete_query => sprintf('DELETE FROM %s WHERE id = ?', $table_name),

        failed_query => sprintf('UPDATE %s SET retry_cnt = retry_cnt + 1, run_after = ?, grabbed_until = 0, priority = ? WHERE id = ?', $table_name),

        unixtime_query => _settled_unixtime_query($driver),

    }, $class;
}

sub _parse_functions {
    my $opts = shift;

    my $functions = $opts->{functions} || [];
    my $default_grab_for = $opts->{default_grab_for} || (60*60);

    my $funcs = +{};
    for (my $i = 0; $i < @{$functions}; $i++) {
        my $func = $functions->[$i];

        my $value;
        if    (not defined $functions->[$i+1]) {$i++                       }
        elsif (ref $functions->[$i+1])         {$value = $functions->[++$i]}

        $value->{grab_for}     ||= $default_grab_for;
        $value->{serializer}   ||= ($opts->{default_serializer}   || sub {$_[0]});
        $value->{deserializer} ||= ($opts->{default_deserializer} || sub {$_[0]});

        $funcs->{$func} = $value;
    }
    $funcs;
}

sub _verify_driver {
    my $dbh = shift;
    my $driver = $dbh->{Driver}{Name};
    $driver =~ /(mysql|SQLite|Pg)/ ? $driver : Carp::croak('Jonk support only mysql,SQLite,Pg');
}

sub _settled_unixtime_query {
    my $driver = shift;

    if ($driver eq 'Pg') {
        return "SELECT TRUNC(EXTRACT('epoch' from NOW()))";
    } elsif ($driver eq 'mysql') {
        return 'SELECT UNIX_TIMESTAMP()';
    }
}

sub errstr {$_[0]->{_errstr}}

sub insert {
    my ($self, $func, $arg, $opt) = @_;

    my $job_id;
    try {
        $self->{_errstr} = undef;
        local $self->{dbh}->{RaiseError} = 1;
        local $self->{dbh}->{PrintError} = 0;

        my $serializer = $self->{functions}->{$func}->{serializer} ||= sub {$_[0]};
        my $sth = $self->{dbh}->prepare_cached($self->{insert_query});
        $sth->bind_param(1, $func);
        $sth->bind_param(2, $serializer->($arg), _bind_param_attr($self->{driver}));
        $sth->bind_param(3, time());
        $sth->bind_param(4, $opt->{run_after}||0);
        $sth->bind_param(5, $opt->{priority} ||0);
        $sth->execute();

        $job_id = $self->{dbh}->last_insert_id("","",$self->{table_name},"");
        $sth->finish;
    } catch {
        $self->{_errstr} = "can't insert for job queue database: $_"
    };

    $job_id;
}

sub _bind_param_attr {
    my $driver = shift;

    if ( $driver eq 'Pg' ) {
        return { pg_type => DBD::Pg::PG_BYTEA() };
    } elsif ( $driver eq 'SQLite' ) {
        return DBI::SQL_BLOB();
    }
    return;
}

sub _server_unixitime {
    my $self = shift;
    return time() if $self->{driver} eq 'SQLite';
    $self->{dbh}->selectrow_array($self->{unixtime_query});
}

sub _grab_job {
    my ($self, $callback, $opt) = @_;

    my $job;
    try {
        $self->{_errstr} = undef;
        local $self->{dbh}->{RaiseError} = 1;
        local $self->{dbh}->{PrintError} = 0;

        my $time = $self->_server_unixitime;
        my $sth = $callback->($time);

        while (my $row = $sth->fetchrow_hashref) {
            $job = $self->_grab_a_job($row, $time, $opt);
            last if $job;
        }

        $sth->finish;
    } catch {
        $self->{_errstr} = "can't grab job from job queue database: $_";
    };

    $job;

}

sub _grab_a_job {
    my ($self, $row, $time) = @_;

    my $sth = $self->{dbh}->prepare_cached($self->{grab_query});
    $sth->execute(
        ($time + ($self->{functions}->{$row->{func}}->{grab_for})),
        $row->{id},
        $row->{grabbed_until}
    );
    my $grabbed = $sth->rows;
    $sth->finish;
    $grabbed ? Jonk::Job->new($self => $row) : undef;
}

sub lookup_job {
    my ($self, $job_id) = @_;

    $self->_grab_job(
        sub {
            my $time = shift;
            my $sth = $self->{dbh}->prepare_cached($self->{lookup_job_query});
            $sth->execute($job_id, $time, $time);
            $sth;
        }
    );
}

sub find_job {
    my $self = shift;

    unless ($self->{has_func}) {
        Carp::croak('missin find_job functions.');
    }

    $self->_grab_job(
        sub {
            my $time = shift;
            my $sth = $self->{dbh}->prepare_cached($self->{find_job_query});
            $sth->execute($time, $time);
            $sth;
        }
    );
}

sub _delete {
    my ($self, $job_id) = @_;

    try {
        my $sth = $self->{dbh}->prepare_cached($self->{delete_query});
        $sth->execute($job_id);
        $sth->finish;
        return $sth->rows;
    } catch {
        $self->{_errstr} = "can't dequeue job from job queue database: $_";
        return;
    };
}

sub _failed {
    my ($self, $job_id, $opt) = @_;

    my $retry_delay = $self->_server_unixitime + (defined($opt->{retry_delay}) ? $opt->{retry_delay} : 60);
    my $priority    = (defined($opt->{priority}) ? $opt->{priority} : 0);

    try {
        my $sth = $self->{dbh}->prepare_cached($self->{failed_query});
        $sth->execute($retry_delay, $priority, $job_id);
        $sth->finish;
        return $sth->rows;
    } catch {
    warn 'ababaaaba';
        $self->{_errstr} = "can't update job from job queue database: $_";
        return;
    };
}

1;

__END__

=head1 NAME

Jonk - simple job tank manager.

=head1 SYNOPSIS

    use DBI; 
    use Jonk;
    my $dbh = DBI->connect(...);
    my $jonk = Jonk->new($dbh, {functions => [qw/MyWorker/]});
    # insert job
    {
        $jonk->insert('MyWorker', 'arg');
    }

    # execute job
    {
        my $job = $jonk->find_job;
        print $job->func; # MyWorker
        print $job->arg;  # arg
        $job->completed;
    }

=head1 DESCRIPTION

Jonk is simple job queue manager system

Job is saved and taken out. Besides, nothing is done.

You may use Jonk to make original Job Queuing System.

=head1 METHODS

=head2 my $jonk = Jonk::Worker->new($dbh, [\%options]);

Creates a new Jonk object, and returns the object.

$options is an optional settings.

=over 4

=item * $dbh

$dbh is database handle.

=item * $options->{functions}

Key word of job which this Jonk instance looks for.

=over 4

=item * $options->{functions} = [qw/worker_key worker_key2/]

can set *worker_key* at arrayref.

=item * $options->{functions} = ['worker_key' => {grab_for => 5}],

can set worker_key's grab_for setting by hash-ref.

=item * $options->{functions} = ['worker_key' => {serializer => \&serialize_code, deserializer => \&deserialize_code}],

can set worker_key's (de)serializer code setting by hash-ref.

=item * $options->{functions} = ['worker_key' => {serializer => \&serialize_code, deserializer => \&deserialize_code}, 'worker_key2'],

can mix worker settings.

=back

=item * $options->{table_name}

specific job table name.

Default job table name is `job`.

=item * $options->{job_find_size}

specific lookup job record size.

Default 50.

=item * $options->{default_serializer}

global serializer setting.

=item * $options->{default_deserializer}

global deserializer setting.

=item * $options->{default_grab_for}

global grab_for setting.

=back

=head2 my $job_id = $jonk->insert($func, $arg);

enqueue a job to a database.
returns job.id.

=over 4

=item * $func

specific your worker funcname.

=item * $arg

job argument data.

=back

=head2 my $job = $jonk->lookup_job($job_id);

lookup a job from a database.

returns Jonk::Job object.

=over 4

=item * $job_id

lookup specific $job_id's job.

=back

=head2 my $job = $jonk->find_job();

get job from database by sorted priority descending order.

=head2 $jonk->errstr;

get most recent error infomation.

=head1 ERROR HANDLING

  my $job = $jonk->lookup;
  if ($jonk->errstr) {
      die $jonk->errstr;
  }

=head1 SCHEMA

=head2 MySQL

    CREATE TABLE job (
        id            int(10) UNSIGNED NOT NULL auto_increment,
        func          varchar(255)     NOT NULL,
        arg           MEDIUMBLOB,
        enqueue_time  INTEGER UNSIGNED,
        grabbed_until int(10) UNSIGNED NOT NULL,
        run_after     int(10) UNSIGNED NOT NULL DEFAULT 0,
        retry_cnt     int(10) UNSIGNED NOT NULL DEFAULT 0,
        priority      int(10) UNSIGNED NOT NULL DEFAULT 0,
        primary key ( id )
    ) ENGINE=InnoDB

=head2 SQLite

    CREATE TABLE job (
        id            INTEGER PRIMARY KEY ,
        func          text,
        arg           text,
        enqueue_time  INTEGER UNSIGNED,
        grabbed_until INTEGER UNSIGNED NOT NULL,
        run_after     INTEGER UNSIGNED NOT NULL DEFAULT 0,
        retry_cnt     INTEGER UNSIGNED NOT NULL DEFAULT 0,
        priority      INTEGER UNSIGNED NOT NULL DEFAULT 0
    )

=head2 PostgreSQL

    CREATE TABLE job (
        id            SERIAL PRIMARY KEY,
        func          TEXT NOT NULL,
        arg           BYTEA,
        enqueue_time  INTEGER,
        grabbed_until INTEGER NOT NULL,
        run_after     INTEGER NOT NULL DEFAULT 0,
        retry_cnt     INTEGER NOT NULL DEFAULT 0,
        priority      INTEGER NOT NULL DEFAULT 0
    )

=head1 SEE ALSO

L<Qudo>

L<TheSchwartz>

=head1 SUPPORT

  irc: #jonk@irc.perl.org

=head1 REPOSITORY

  git clone git://github.com/nekokak/p5-Jonk.git

=head1 CONTRIBUTORS

tokuhirom

kan_fushihara

fujiwara

=head1 AUTHOR

Atsushi Kobayashi E<lt>nekokak _at_ gmail _dot_ comE<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

